x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -0,0 +1,418 @@
+import {
+  ActionsCompletion,
+  AlertInstanceContext,
+  AlertInstanceState,
+  RuleAction,
+  RuleTypeParams,
+  RuleTypeState,
+  SanitizedRule,
+} from '../../common';
+
+export class ExecutionHandler<
+  Params extends RuleTypeParams,
+  ExtractedParams extends RuleTypeParams,
+  RuleState extends RuleTypeState,
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+> {
+  public request: KibanaRequest;
+  private logger: Logger;
+  private alertingEventLogger: PublicMethodsOf<AlertingEventLogger>;
+  private rule: SanitizedRule<Params>;
+  private ruleType: NormalizedRuleType<
+    Params,
+    ExtractedParams,
+    RuleState,
+    State,
+    Context,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >;
+  private ruleTypeActionGroups;
+  private taskRunnerContext: TaskRunnerContext;
+  private taskInstance: RuleTaskInstance;
+  private ruleRunMetricsStore: RuleRunMetricsStore;
+  private apiKey: RawRule['apiKey'];
+  private ruleConsumer: string;
+  private executionId: string;
+  private ruleLabel: string;
+  private ephemeralActionsToSchedule: number;
+  private mutedAlertIdsSet: Set<string>;
+  private CHUNK_SIZE = 1000;
+  private skippedAlerts: { [key: string]: { reason: string } } = {};
+
+  constructor({
+    rule,
+    ruleType,
+    logger,
+    alertingEventLogger,
+    taskRunnerContext,
+    taskInstance,
+    ruleRunMetricsStore,
+    apiKey,
+    ruleConsumer,
+    executionId,
+    ruleLabel,
+    request,
+  }: ExecutionHandlerOptions<
+    Params,
+    ExtractedParams,
+    RuleState,
+    State,
+    Context,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >) {
+    this.logger = logger;
+    this.alertingEventLogger = alertingEventLogger;
+    this.rule = rule;
+    this.ruleType = ruleType;
+    this.ruleTypeActionGroups = new Map(
+      ruleType.actionGroups.map((actionGroup) => [actionGroup.id, actionGroup.name])
+    );
+    this.taskRunnerContext = taskRunnerContext;
+    this.taskInstance = taskInstance;
+    this.ruleRunMetricsStore = ruleRunMetricsStore;
+    this.apiKey = apiKey;
+    this.ruleConsumer = ruleConsumer;
+    this.executionId = executionId;
+    this.ruleLabel = ruleLabel;
+    this.request = request;
+    this.ephemeralActionsToSchedule = taskRunnerContext.maxEphemeralActionsPerRule;
+    this.mutedAlertIdsSet = new Set(rule.mutedInstanceIds);
+  }
+
+  public async run(
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>,
+    recovered: boolean = false
+  ) {
+    const {
+      CHUNK_SIZE,
+      logger,
+      alertingEventLogger,
+      ruleRunMetricsStore,
+      taskRunnerContext: { actionsConfigMap, actionsPlugin },
+      taskInstance: {
+        params: { spaceId, alertId: ruleId },
+      },
+    } = this;
+
+    const executables = this.generateExecutables({ alerts, actions: this.rule.actions, recovered });
```

Comment:
nit: seems like we don't need to pass `this.rule.actions` into `generateExecutables` since that function belongs to this class and should have access to it already?